improver/psychrometric_calculations/psychrometric_calculations.py

```diff
@@ -315,23 +319,62 @@ def _make_humidity_cube(self, data: np.ndarray) -> Cube:
         )
         return cube
 
+    def generate_pressure_cube(self) -> None:
+        """Generate a pressure cube from the pressure coordinate on the temperature cube"""
+        coord_list = [coord.name() for coord in self.temperature.coords()]
+        pressure_list = CubeList()
+        for temp_slice in self.temperature.slices_over("pressure"):
+            pressure_value = temp_slice.coord("pressure").points
+            temp_slice.data = np.broadcast_to(pressure_value, temp_slice.shape)
+            pressure_list.append(temp_slice)
+        self.pressure = pressure_list.merge_cube()
+        enforce_coordinate_ordering(self.pressure, coord_list)
+        self.pressure.rename("surface_air_pressure")
+        self.pressure.units = "Pa"
+
     def process(self, *cubes: Union[Cube, CubeList]) -> Cube:
         """
-        Calculates the humidity mixing ratio from the inputs.
+        Calculates the humidity mixing ratio from the inputs. The inputs can be on height levels
+        or pressure levels, and the output will be on the same levels. If the input cubes are on
+        pressure levels then a pressure cube doesn't need to be provided and the pressure levels
+        will be inferred from the pressure coordinate on the temperature cube.
 
         Args:
             cubes:
-                Cubes of temperature (K), pressure (Pa) and relative humidity (1)
+                Cubes of temperature (K) and relative humidity (1). A cube of pressure (Pa) must also
+                br provided unless there is a pressure coordinate in the temperature and relative humidity cubes.
 
         Returns:
-            Cube of humidity mixing ratio
+            Cube of humidity mixing ratio on same levels as input cubes
 
         """
         cubes = as_cubelist(*cubes)
-        (self.temperature, self.pressure, self.rel_humidity) = cubes.extract_cubes(
-            ["air_temperature", "surface_air_pressure", "relative_humidity"]
+
+        (self.temperature, self.rel_humidity) = cubes.extract_cubes(
+            ["air_temperature", "relative_humidity"]
         )
 
+        try:
+            self.pressure = cubes.extract_cube("surface_air_pressure")
+        except ConstraintMismatchError:
+            # If no pressure cube is provided, check if pressure is a coordinate in the temperature and relative humidity cubes
+            temp_coord = [
```

Comment:
This section seemed a little confusing to me. This is almost exactly the same thing but might be easier to understand/a tiny tiny bit faster:
```python
temp_coord_flag = any(
    coord.name() == "pressure"
    for coord in self.temperature.coords()
)
rh_coord_flag = any(
    coord.name() == "pressure"
    for coord in self.rel_humidity.coords()
)
if temp_coord_flag & rh_coord_flag:
    self.generate_pressure_cube()
else:
    raise ValueError(
        "No pressure cube called 'surface_air_pressure' found and no pressure coordinate found in temperature or relative humidity cubes"
        )

Reply:
I prefer your code to mine so I've updated it.